src/fs/cluster_allocator.cc

```diff
@@ -21,34 +21,71 @@
 
 #include "fs/cluster.hh"
 #include "fs/cluster_allocator.hh"
+#include "seastar/core/future.hh"
 
 #include <cassert>
 #include <optional>
 
-namespace seastar {
+namespace seastar::fs {
 
-namespace fs {
+cluster_allocator::cluster_allocator() : _cluster_sem(0) {}
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
+void cluster_allocator::init(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters) {
+    _allocated_clusters = std::move(allocated_clusters);
+    _free_clusters = std::move(free_clusters);
+    assert(_cluster_sem.waiters() == 0);
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
+}
 
-std::optional<cluster_id_t> cluster_allocator::alloc() {
-    if (_free_clusters.empty()) {
-        return std::nullopt;
-    }
+cluster_id_t cluster_allocator::do_alloc() {
+    assert(not _free_clusters.empty());
 
     cluster_id_t cluster_id = _free_clusters.front();
     _free_clusters.pop_front();
     _allocated_clusters.insert(cluster_id);
+
     return cluster_id;
 }
 
-void cluster_allocator::free(cluster_id_t cluster_id) {
+cluster_id_t cluster_allocator::do_free(cluster_id_t cluster_id) {
```

Comment:
1. returning `cluster_id_t` looks pointless.
2. A comment could indicate that the strong exception guarantee is provided